File: mesmaeker-mockito/src/main/java/fr/xgouchet/mesmaeker/mockito/ValidatingOngoingReturningClause.java
package fr.xgouchet.mesmaeker.mockito;

import java.util.function.Function;

import static org.assertj.core.api.Java6Assertions.assertThat;

/**
 * @author Xavier F. Gouchet
 */
public class ValidatingOngoingReturningClause<T, O> implements OngoingReturningClause<T, O> {

    private final T subject;
    private final O firstResult;
    private final O[] nextResults;

    @SuppressWarnings("unchecked")
    public ValidatingOngoingReturningClause(T subject, O firstResult, O... nextResults) {
        this.subject = subject;
        this.firstResult = firstResult;
        this.nextResults = nextResults;
    }

    @Override
    public OngoingReturningClause<T, O> when(Function<T, O> function) {
        assertThat(function.apply(subject)).isEqualTo(firstResult);

        if (nextResults != null) {
            for (O nextResult : nextResults) {
                assertThat(function.apply(subject)).isEqualTo(nextResult);
            }
        }
        return this;
    }
}

File: mesmaeker-mockito/src/main/java/fr/xgouchet/mesmaeker/mockito/ValidatingOngoingReturningClause.kt
package fr.xgouchet.mesmaeker.mockito

import java.util.function.Function


/**
 * @author Xavier F. Gouchet
 */
class ValidatingOngoingReturningClause<T, O>(private val subject: T,
                                             private val value: O,
                                             vararg values: O)
    : OngoingReturningClause<T, O> {

    private val nextValues: List<O> = listOf(*values)

    override fun `when`(function: Function<T, O>): OngoingReturningClause<T, O> {
        return whenever(function)
    }

    override fun whenever(function: Function<T, O>): OngoingReturningClause<T, O> {
        var actual = function.apply(subject)
        require(actual == value, { "Expected <$actual> to be <$value>" })

        for (nextValue in nextValues) {
            actual = function.apply(subject)
            require(actual == nextValue, { "Expected <$actual> to be <$nextValue>" })
        }
        return this
    }
}
